### ISSUE TITLE
PHP Parser can't handle bitwise operators

### ISSUE THREAD

**Issue by nevadascout** [HvyIndustries/crane#85]
Investigate these stub errors:
```
78 Stub Error: /Users/ns/Library/Preferences/Crane/phpstubs/crane-php-stubs-master/standard.php
74 Stub Error: /Users/ns/Library/Preferences/Crane/phpstubs/crane-php-stubs-master/SPL.php
6 Stub Error: /Users/ns/Library/Preferences/Crane/phpstubs/crane-php-stubs-master/Core.php
```

It seems that this function is causing the issue:
```php
function htmlspecialchars(
    string $string,
    int $flags = ENT_COMPAT | ENT_HTML401,
    string $encoding = 'ini_get("default_charset")',
    bool $double_encode = true
): string {}
```
`ENT_COMPAT | ENT_HTML401` is breaking it -- need to file a bug report with the php-parser project regarding bitwise operators.

Worth investigating if the `ini_get` section is valid PHP...

**Comment by nevadascout** [comment]
I seem to still be getting an error when trying to parse those three stubs on the newest version of the parser, need to investigate further...

**Comment by ichiriac** [comment]
Strange, it's like you have some cache and can not retrieve the latest version. 

I've got the same behavior here : https://tonicdev.com/ichiriac/php-parser-function

I've released the version 0.0.10 on npm, it will maybe help.

Try to run this from your console :

```bash
Z:\php-parser\bin>node test.js --debug -e "function htmlspecialchars(string $string, int $flags = ENT_COMPAT | ENT_HTML401, string $encoding = 'ini_get(""default_charset"")', bool $double_encode = true ): string {}"

*** START TESTING ***

-- TOKENS :
T_FUNCTION T_STRING ( T_STRING T_VARIABLE , T_STRING T_VARIABLE = T_STRING | T_STRING , T_STRING T_VARIABLE = T_CONSTANT_ENCAPSED_STRING , T_STRING T_VARIABLE = T_STRING ) : T_STRING { }
Line 1 : T_FUNCTION @at Object.api.parse (Z:\php-parser\src\parser.js:185:12)
Line 1 : T_STRING @at Object.read_function_declaration (Z:\php-parser\src\parser\function.js:58:24)
Line 1 : '(' @at Object.read_function_declaration (Z:\php-parser\src\parser\function.js:62:14)
Line 1 : T_STRING @at Object.read_function_declaration (Z:\php-parser\src\parser\function.js:64:24)
Line 1 : T_VARIABLE @at Object.api.read_list (Z:\php-parser\src\parser.js:382:21)
Line 1 : ',' @at Object.read_parameter (Z:\php-parser\src\parser\function.js:129:16)
Line 1 : T_STRING @at Object.read_parameter_list (Z:\php-parser\src\parser\function.js:107:18)
Line 1 : T_VARIABLE @at Object.api.read_list (Z:\php-parser\src\parser.js:382:21)
Line 1 : '=' @at Object.read_parameter (Z:\php-parser\src\parser\function.js:129:16)
Line 1 : T_STRING @at Object.read_parameter (Z:\php-parser\src\parser\function.js:130:22)
Line 1 : '|' @at Object.api.read_list (Z:\php-parser\src\parser.js:382:21)
Line 1 : T_STRING @at Object.read_expr (Z:\php-parser\src\parser\expr.js:15:59)
Line 1 : ',' @at Object.api.read_list (Z:\php-parser\src\parser.js:382:21)
Line 1 : T_STRING @at Object.read_parameter_list (Z:\php-parser\src\parser\function.js:107:18)
Line 1 : T_VARIABLE @at Object.api.read_list (Z:\php-parser\src\parser.js:382:21)
Line 1 : '=' @at Object.read_parameter (Z:\php-parser\src\parser\function.js:129:16)
Line 1 : T_CONSTANT_ENCAPSED_STRING @at Object.read_parameter (Z:\php-parser\src\parser\function.js:130:22)
Line 1 : ',' @at Object.read_scalar (Z:\php-parser\src\parser\scalar.js:40:18)
Line 1 : T_STRING @at Object.read_parameter_list (Z:\php-parser\src\parser\function.js:107:18)
Line 1 : T_VARIABLE @at Object.api.read_list (Z:\php-parser\src\parser.js:382:21)
Line 1 : '=' @at Object.read_parameter (Z:\php-parser\src\parser\function.js:129:16)
Line 1 : T_STRING @at Object.read_parameter (Z:\php-parser\src\parser\function.js:130:22)
Line 1 : ')' @at Object.api.read_list (Z:\php-parser\src\parser.js:382:21)
Line 1 : ':' @at Object.read_function_declaration (Z:\php-parser\src\parser\function.js:66:24)
Line 1 : T_STRING @at Object.read_function_declaration (Z:\php-parser\src\parser\function.js:72:27)
Line 1 : '{' @at Object.api.read_list (Z:\php-parser\src\parser.js:382:21)
Line 1 : '}' @at Object.read_code_block (Z:\php-parser\src\parser\statement.js:253:24)
Line 1 : the end of file (EOF) @at Object.read_code_block (Z:\php-parser\src\parser\statement.js:258:24)
-- AST :
[ 'program',
  [ [ 'position',
      [ 1, 0, 0 ],
      [ 1, 168, 169 ],
      [ 'function',
        'htmlspecialchars',
        [ [ '$string', [ 'string' ], [], false, false ],
          [ '$flags',
            [ 'int' ],
            [ 'position',
              [ 1, 66, 66 ],
              [ 1, 68, 79 ],
              [ 'bin',
                '|',
                [ 'const', 'ENT_COMPAT' ],
                [ 'const', 'ENT_HTML401' ] ] ],
            false,
            false ],
          [ '$encoding',
            [ 'string' ],
            [ 'string', 'ini_get("default_charset")' ],
            false,
            false ],
          [ '$double_encode', [ 'bool' ], [ 'const', 'true' ], false, false ] ],
        false,
        [],
        [ 'string' ],
        [] ] ] ] ]

```

**Comment by TheColorRed** [comment]
Could you test it on this file? It is possible that their are other things that are breaking parser

https://github.com/HvyIndustries/crane-php-stubs/blob/master/standard.php

**Comment by ichiriac** [comment]
it's passing :

```bash
# the quick test way : node test.js -f ../test/token/php7-stubs.php

Z:\php-parser\bin>node test.js -f ../test/token/php7-stubs.php --debug
.... many lines ...
    [ 'call',
      [ 'ns', [ 'define' ] ],
      [ [ 'string', 'DNS_ALL' ], [ 'number', 251713587 ] ] ],
    null ] ]
v - Passed AST parsing
Success
```

**Comment by ichiriac** [comment]
by the way, now it's also ok on tonicdev with the new version flag : https://tonicdev.com/ichiriac/php-parser-function

**Comment by nevadascout** [comment]
You were exactly right @ichiriac - it was caching the old version.

Fix was to terminate the build task on the server and start it again - that updates the server npm packages into the client folder. That might be a potential fix for you too @TheColorRed for that other issue.

**Comment by nevadascout** [comment]
This is now resolved in 07f2fc077743965881f557e28eb3928006eb8543
